tensorflow/core/common_runtime/zen_layout_pass.cc

```diff
@@ -0,0 +1,1196 @@
+    // inference. Currently Zen supports inference in float only.
+    if (!CheckValidityForDTypeSupported(n)) {
+      return false;
+    }
+    std::vector<string> fused_ops;
+    TF_CHECK_OK(GetNodeAttr(n->def(), "fused_ops", &fused_ops));
+
+    return (fused_ops == std::vector<string>{"BiasAdd"} ||
+            fused_ops == std::vector<string>{"FusedBatchNorm"} ||
+            fused_ops == std::vector<string>{"Relu"} ||
+            fused_ops == std::vector<string>{"BiasAdd", "Relu"} ||
+            fused_ops == std::vector<string>{"BiasAdd", "Relu6"} ||
+            fused_ops == std::vector<string>{"BiasAdd", "Add"} ||
+            fused_ops == std::vector<string>{"BiasAdd", "Add", "Relu"} ||
+            fused_ops == std::vector<string>{"FusedBatchNorm", "Relu"});
+  }
+
+  // Currently TF-ZenDNN supports FP32 inference only. Returns, true if node is
+  // of float dataype, false otherwise.
+  static bool CheckValidityForDTypeSupported(const Node *n) {
+    DataType data_type;
+    TF_CHECK_OK(GetNodeAttr(n->def(), "T", &data_type));
+    return (data_type == DT_FLOAT);
+  }
+
+  // Method to provide a 'valid' status for nodes that don't require any check.
+  // This method is used in ZenLayoutRewritePass() for creating the record/entry
+  // for rewriting native ops with Zen ops.
+  static bool RewriteValid(const Node *n) { return true; }
+
+  // Method to find whether the graph has inference ops only. It returns error
+  // status if the graph has training ops.
+  Status AreAllInferenceOps(std::unique_ptr<Graph> *g);
+
+  // Rewrites input node to a new node specified by its matching rewrite record.
+  //
+  // Input node may be deleted in case of rewrite. Attempt to use the node
+  // after the call can result in undefined behaviors.
+  //
+  // @input  g - input graph, n - Node to be rewritten,
+  //         ri - matching rewrite record,
+  //         reorder_flags - flags to populate reorder attributes of Zen op.
+  // @return OkStatus(), if the input node is rewritten;
+  //         Returns appropriate Status error code otherwise.
+  //         Graph is updated in case the input node is rewritten.
+  //         Otherwise, it is not updated.
+  Status ZenOpNodeRewrite(std::unique_ptr<Graph> *g, Node *n,
+                          const ZenOpRewriteRecord *ri,
+                          std::pair<bool, bool> reorder_flags);
+
+  // Functions specific to operators to copy attributes
+  // We need operator-specific function to copy attributes because the framework
+  // does not provide any generic function for it.
+  static void UpdateZenOpAttrs(const Node *orig_node, NodeBuilder *nb);
+
+  static void UpdateZenOpAttrsConv2D(const Node *orig_node, NodeBuilder *nb);
+
+  static void UpdateZenOpAttrsFusedConv2D(const Node *orig_node,
+                                          NodeBuilder *nb);
+
+  // Examines the input and output nodes of each node, for Zen nodes and
+  // determines reorder flags.
```

Comment:
Please explain that reorder here means converting tensor layout (otherwise people could confuse it with node reordering), and that reorder_before and reorder_after indicates whether the tensors need to be reordered to Zen format before / after the Zen node.

Reply:
We have added necessary explanation for reorder flags. Thank you!